Fix Replication Factor calculation for Tablets effective replication map.

bitpathfinder opened the pull request:
Current implementation of replication factor calculation does not take into account Tablets migration. Since a table can be represented by multiple Tablets, during Tablets migration there are different Tablets could be in different states, thus there cannot be a single state for the table and previously used logic of taking the RF from the schema as it worked for VNodes, is not valid when tablets are in a migration state.

In this PR, the `effective_replication_map::get_replication_factor()` implementation was modified for Tablets. Instead of retrieving the replication factor from the schema, it now calculates the replication factor based on the actual replica set from the Tablets map. This change addresses potential inconsistencies between the schema and the state of individual tablets during migration.

Fixes scylladb/scylladb#20625 
Fixes scylladb/scylladb#20282

Since this fixes an important issue with Tablets, we need backports to 6.0, 6.1, 6.2.

bitpathfinder committed:
Fix error message 'Invalid replica selector' format.

bitpathfinder committed:
storage_proxy: Coroutinize do_make_replication_map.

Since the construction of the tablet_effective_replication_map will
use `maybe_yield()` to avoid reactor stalls `do_make_replication_map()`
needs to be converted into a coroutine.

bitpathfinder committed:
Rename `get_replication_factor` to `get_schema_replication_factor`

As part of introducing a new method for calculating the replication
factor for Tablets, rename the existing `get_replication_factor` to
`get_schema_replication_factor`, since it returns the replication
factor from the replication strategy, whereas the new method will
calculate the replication factor based on the current state of the
Tablets.

bitpathfinder committed:
Refactor `tablet_effective_replication_map::get_replica*` functions.

1. Split the functionality of get_endpoints_for_reading into two functions:

- `get_endpoints_for_reading(const tablet_id tablet)` returns the current
   read replica host ID set from Tablets. This change is necessary as part
   of introducing new replication factor calculation functions.
- `get_endpoints_for_reading(const token& search_token)` uses the new function
   that inputs a Tablet id.

2. Move `get_endpoints_for_reading`, `get_pending_endpoints`, and `get_replicas_for_write`
   logic for finding replicas in the Tablet map to the `tablet_map` class.
   This change is needed to build the replication factor map outside of the ERM
   and improves encapsulation of the `tablet_map` logic.

bitpathfinder committed:
Add `effective_replication_map::get_replication_factor` functions.

Add implementations for new get_replication_factor functions in the
effective replication map that for Tablets calculate the replication
factor based on actual replica counts from the Tablets map. This accounts for
potential inconsistencies between the schema and Tablets during migration.

bitpathfinder committed:
Modify storage_proxy to use the tablet-aware `get_replication_factor`.

Replace `get_schema_replication_factor` with the new `get_replication_factor`,
which, for tablets, returns the actual number of replicas while accounting
for the transition state of Tablets.

Fixes scylladb/scylladb#20625

bitpathfinder committed:
Enable test_change_replication_factor_1_to_0 test for Tablets.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12791/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12791/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_change_replication_factor_1_to_0
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12791/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1381/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12472/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4450/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12791/)
#### Build Details:
- Duration: 7 hr 44 min
- Builder: i-07177465d1a75ff45 (m5ad.8xlarge)

tgrabiec commented on locator/abstract_replication_strategy.cc:
```diff
@@ -558,6 +558,18 @@ vnode_effective_replication_map::factory_key vnode_effective_replication_map::ma
+    if (replication_strategy.get_type() == replication_strategy_type::network_topology) {
+        const network_topology_strategy* nrs = static_cast<const network_topology_strategy*>(&replication_strategy);
+        return nrs->get_replication_factor(datacenter);
+    }
```
Would be cleaner to add dc-aware get_replication_factor() to abstract_replication_strategy so that abstract things stay abstract, and this does not have to be aware of every impl.

tgrabiec replied:
On the other hand, it mimics preexisting logic, so you can ignore.

tgrabiec commented on locator/tablets.cc:
```diff
@@ -616,9 +616,19 @@ future<bool> check_tablet_replica_shards(const tablet_metadata& tm, host_id this
     mutable const tablet_map* _tmap = nullptr;
+    // The `_datacenter_map` stores the association between each data center
+    // and its index in the `replication_factor_list` in `_tablet_to_dc_replication_factor_map`.
+    absl::flat_hash_map<sstring, replication_factor_t> _datacenter_map;
```
If it's an index, it shouldn't be replication_factor_t, but some other type (e.g. size_t). It's not replication factor which is stored.

bitpathfinder replied:
Yes, actually it's a replication factor index type, so I will rename it accordingly.

tgrabiec commented on locator/tablets.cc:
```diff
@@ -616,9 +616,19 @@ future<bool> check_tablet_replica_shards(const tablet_metadata& tm, host_id this
+    absl::flat_hash_map<sstring, replication_factor_t> _datacenter_map;
+    // The `_tablet_to_dc_replication_factor_map` stores replication factors
+    // per data center for each Tablet.
+    std::vector<replication_factor_list> _tablet_to_dc_replication_factor_map;
```
Should be chunked_vector to avoid large contiguous allocations.

bitpathfinder replied:
Can you please elaborate why is it bad to have a large allocation? Do you mean our allocator's storage might be already so fragmented that there would not be such a big chunk of memory (~5 Mb for 100k tablets ) ?

tgrabiec replied:
Yes, the larger the allocation to more likely it is to cause premature OOM. In practice, we will try to reclaim contiguous space back from the LSA region (memtables/cache), which occupies higher addresses, but that may still require a lot of work to evict data and move segments around, and may cause reactor stalls. We maintain a reserve of free standard allocator space which is supposed to amortizes this cost, but if the request cannot be met due to fragmentation, it cannot make use of that reserve. 

As a general rule, we avoid contiguous allocations larger than 128K (LSA segment size). 128K is a unit of memory reclamation from LSA to the standard allocator's pool.

bitpathfinder replied:
In the benchmark I see that the number of tablets per table is 2048. Is that a typical number for production or it's just a testing number? Also when I set it to 65k I get an error "Exiting on unhandled exception: std::bad_alloc".

bitpathfinder replied:
@tgrabiec  So I suppose the max Tablets per table will be 8192, since otherwise our own tests can't be run. In that case the memory allocated by the `_tablet_to_dc_replication_factor_map` vector is 8192*8=65536 bytes (counting there are 4 data centers). Do you still think it makes sense to switch to a chunked vector (which allocates by default by 128k).

kbr-scylla replied:
>  So I suppose the max Tablets per table will be 8192

I think it's an order of magnitude or two below what we're aiming for.

Compare to vnodes -- the number of vnodes in 100 node cluster is by default 25600 (256 vnodes per node), then multiply it by number of shards per node, e.g. 64 -- becomes ~1.6*10^6  and I imagine we can reach similar number of tablets. More if cluster becomes larger.

tgrabiec replied:
> In the benchmark I see that the number of tablets per table is 2048. Is that a typical number for production or it's just a testing number? Also when I set it to 65k I get an error "Exiting on unhandled exception: std::bad_alloc".

Can you share the command line which fails?

We aim at 100k tablets per cluster.

Some tests may not be able to handle that in a single shard due to overhead of tablet replica. In a real cluster, that overhead would be spread among many shards.

bitpathfinder replied:
@tgrabiec I run the following cmd 
`./build/dev/scylla perf-tablets  --tablets-per-table 65536`  
It fails with `std::bad_alloc`. In fact everything higher than 8192 is failing.

tgrabiec replied:
> @tgrabiec I run the following cmd `./build/dev/scylla perf-tablets --tablets-per-table 65536` It fails with `std::bad_alloc`. In fact everything higher than 8192 is failing.

The default number of tables is 100, so with --tablets-per-table 65536 you have 6.5 M tablets. That's beyond our target of 100k.

tablet_map takes 600 MB. It probably bad_allocs due to mutation object size when persisting. With -m8G it proceeds, but then it takes almost a minute to save:

```
INFO  2024-11-06 14:06:48,099 [shard 0:stat] testlog - Saved in 41656.312500 [ms]
INFO  2024-11-06 14:07:05,748 [shard 0:stat] testlog - Read in 17534.886719 [ms]
```

And we fail in read_tablet_mutations() due to:

```
ERROR 2024-11-06 14:07:08,778 [shard 0:main] seastar - Exiting on unhandled exception: std::runtime_error (Maximum amount of memory for building query results is exhausted, unpaged query cannot be finished)
```

That one is used in migration_manager.cc, in schema pull handler:
```
            if (options->group0_snapshot_transfer) {
                cm.emplace_back(co_await db::system_keyspace::get_group0_history(db));
                if (proxy.local().local_db().get_config().enable_tablets()) {
                    for (auto&& m: co_await replica::read_tablet_mutations(db)) {
```

We shouldn't use that path anymore for snapshot transfer, right @kbr-scylla ?

tgrabiec replied:
I opened https://github.com/scylladb/scylladb/issues/21456

tgrabiec commented on locator/tablets.cc:
```diff
@@ -700,6 +722,71 @@ class tablet_effective_replication_map : public effective_replication_map {
+                const sstring& datacenter = topo.get_datacenter(node.host);
+                const auto it = _datacenter_map.find(datacenter);
+                if (it == _datacenter_map.end()) {
+                    tablet_logger.debug("Could not find datacenter: {} for node: {} ", datacenter, node);
```
That's a rather serious error because it will result in an inconsistency between replication factor and read replica set. So the problem the patch tries to fix. And this one if happens, may be permanent. Probably deserves on_internal_error().

bitpathfinder replied:
But sometimes this is the case, due to a short time of inconsistency between the Tablets and the schema.

kbr-scylla replied:
But this code is performing lookup in the `locator::topology` data structure. This data structure is completely independent from any schema definitions. It's based purely on topology state. It should never be inconsistent with tablets.

Am I missing something?

bitpathfinder replied:
I did not investigate much, but I was getting these errors in the tests consistently, when ERM was created before the node was initialized.

tgrabiec commented on locator/tablets.cc:
```diff
@@ -700,6 +722,71 @@ class tablet_effective_replication_map : public effective_replication_map {
+                    continue;
+                }
+                const sstring& datacenter = topo.get_datacenter(node.host);
+                const auto it = _datacenter_map.find(datacenter);
```
Does erm construction time significantly increases because of this for a large tablet count (100k)?

If so, this could be optimized by keeping dc index in topology.

tgrabiec commented on locator/tablets.cc:
```diff
@@ -700,6 +722,71 @@ class tablet_effective_replication_map : public effective_replication_map {
+            }
+            co_await seastar::maybe_yield();
+        }
+        tablet_logger.info("Added: [{}] records to rf map", _tablet_to_dc_replication_factor_map.size());
```
info is too high for this.

tgrabiec commented on locator/tablets.cc:
```diff
@@ -700,6 +722,71 @@ class tablet_effective_replication_map : public effective_replication_map {
+    [[nodiscard]] size_t get_replication_factor(const dht::token id, const seastar::sstring& datacenter) const override {
+        tablet_logger.trace("Get RF for token {}, dc {}", id, datacenter);
+        auto tablet = get_tablet_map().get_tablet_id(id);
+        assert(tablet.id < _tablet_to_dc_replication_factor_map.size());
```
We prefer on_internal_error() to assert() to not cause unnecessary cluster-wide outages.

tgrabiec commented on locator/tablets.cc:
```diff
@@ -700,6 +722,71 @@ class tablet_effective_replication_map : public effective_replication_map {
+        const auto dc_it = _datacenter_map.find(datacenter);
+        assert(dc_it != _datacenter_map.end());
+        assert(dc_it->second < dc_rf_map.size());
+        if (dc_it == _datacenter_map.end()) [[unlikely]] {
```
It was asserted earlier that it's !=

bitpathfinder replied:
Agree, asserts are redundant. These were useful when I was debugging the code.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -3143,18 +3143,19 @@ future<> storage_service::replicate_to_all_cores(mutable_token_metadata_ptr tmpt
                 auto rs = db.find_keyspace(table->schema()->ks_name()).get_replication_strategy_ptr();
                 locator::effective_replication_map_ptr erm;
                 if (auto pt_rs = rs->maybe_as_per_table()) {
-                    erm = pt_rs->make_replication_map(id, tmptr);
```
Same as above -- please check carefully if this yield won't cause problems

kbr-scylla commented on locator/everywhere_replication_strategy.cc:
```diff
@@ -41,6 +41,11 @@ void everywhere_replication_strategy::validate_options(const gms::feature_servic
 
 sstring everywhere_replication_strategy::sanity_check_read_replicas(const effective_replication_map& erm, const inet_address_vector_replica_set& read_replicas) const {
+    const auto& replication_strategy = erm.get_replication_strategy();
+    if (replication_strategy.uses_tablets()) {
```
We're inside `everywhere_replication_strategy` code, so `erm.get_replication_strategy()` is this strategy, no? In which case this branch is dead code (only NTS can use tablets)

kbr-scylla commented on locator/network_topology_strategy.cc:
```diff
@@ -557,6 +557,12 @@ tablet_replica_set network_topology_strategy::drop_tablets_in_dc(schema_ptr s, c
                                                               const inet_address_vector_replica_set& read_replicas) const {
+    const auto& replication_strategy = erm.get_replication_strategy();
+    if (replication_strategy.uses_tablets()) {
+        // Does not make sense since Tablets are calculating the replication factor basing on the replica set.
```
It doesn't make sense, or it's redundant? Even if it's no longer reproducing the bug (as it got fixed), it still acts nicely to prevent regressions I think.

bitpathfinder replied:
Ok. I will need to fix this check then, since it does not take the token in to account.

gleb-cloudius commented on locator/tablets.cc:
```diff
@@ -867,8 +954,9 @@ std::unordered_set<sstring> tablet_aware_replication_strategy::recognized_tablet
-    return make_ready_future<effective_replication_map_ptr>(
-            seastar::make_shared<tablet_effective_replication_map>(table, std::move(rs), std::move(tm), replication_factor));
+    auto erm = seastar::make_shared<tablet_effective_replication_map>(table, std::move(rs), std::move(tm), replication_factor);
+    co_await erm->build_dc_replication_factor_map();
```
Can we detect that there is no tablets in transition and do not populate anything? If the array is not populated we return values according to RF.

kbr-scylla replied:
IIRC this won't work because when applying `mixed_change` or snapshot there is a short period of time where the schema is updated, but topology is not updated yet. So we construct e_r_m based on updated schema (e.g. RF reduced from 1 to 0) but old topology (tablets are still not in transition state). This state in practice holds for less than a second -- it is basically happening between these two co_awaits:
```cpp
    [&] (mixed_change& chng) -> future<> {
        co_await _mm.merge_schema_from(netw::messaging_service::msg_addr(std::move(cmd.creator_addr)), std::move(chng.mutations));
        co_await _ss.topology_transition();
```

but it's still happening and if we allowed it, the bug is still present.

tgrabiec replied:
If the goal is to reduce lookups and memory in the case of shared RF, we could compute the shared RF map from erm if we notice that all tablets have the same replica count.

kbr-scylla replied:
Either way we should probably fix `mixed_change` so it reloads topology and schema in isolated fashion. (I believe @nuivall's work will address it)

bitpathfinder commented on replica/database.cc:
```diff
@@ -921,7 +921,7 @@ future<> database::add_column_family(keyspace& ks, schema_ptr schema, column_fam
     locator::effective_replication_map_ptr erm;
     if (auto pt_rs = rs.maybe_as_per_table()) {
-        erm = pt_rs->make_replication_map(schema->id(), _shared_token_metadata.get());
+        erm = co_await pt_rs->make_replication_map(schema->id(), _shared_token_metadata.get());
```
After performing some benchmarks, I decided to remove the preemption point from the `build_dc_replication_factor_map` since with the standard testing parameters of 2048 tablets per table, it takes a half of a microsecond).

bitpathfinder replied:
It turns out my previous benchmark test was incorrect as it was measuring tables with 32 Tablets. I have used `perf-load-balancing` to test the actual time, and for 131072 Tablets the replication factor map building time would be about 9.5ms.  So preemption point should be kept.

bitpathfinder commented on locator/tablets.cc:
```diff
@@ -700,6 +722,71 @@ class tablet_effective_replication_map : public effective_replication_map {
+            const tablet_replica_set& replicas = get_endpoints_for_reading(tablet_id{tablet_index});
+            _tablet_to_dc_replication_factor_map.emplace_back();
+            auto& token_rf_list = _tablet_to_dc_replication_factor_map.back();
+            token_rf_list.resize(_datacenter_map.size(), 0);
```
There is no such constructor in `small_vector` :(.

bitpathfinder replied:
But I can apply a trick - to use a dummy source so it will use a copy constructor. Not sure tho if it's any faster.

kbr-scylla replied:
My suggestion was mostly stylistic one. It's not mandatory

But how about adding the missing constructor to `small_vector`? There is such a constructor in `std::vector` so having similar one in `small_vector` makes sense.

bitpathfinder replied:
Actually, after performing some tests, I was going to switch from small vector to `absl::inlined_vector`, since for small things it's more economic in terms of memory. Our `small_vector` has an overhead of 24 bytes just for internal pointers. Btw I believe if we replace it for all our small vector replicas, we could get a notifiable memory consumption decrease (as well as better performance).

bitpathfinder replied:
Added a new constructor to small_vector.

bitpathfinder committed:
Add fill constructor to `small_vector`.

Implement a constructor that initializes `small_vector` with `n` elements, each set to a specified value.

bitpathfinder commented on locator/tablets.cc:
```diff
@@ -700,6 +722,71 @@ class tablet_effective_replication_map : public effective_replication_map {
+    // For each tablet index in the tablet map, it retrieves the current replica set using `get_endpoints_for_reading` 
+    // and counts replicas per data center.  Each data center's replication factor (RF) is saved under 
+    // its index according to `_datacenter_map`.
+    seastar::future<> build_dc_replication_factor_map() {
```
removed

bitpathfinder commented on locator/tablets.cc:
```diff
@@ -690,6 +700,18 @@ class tablet_effective_replication_map : public effective_replication_map {
         return replicas;
     }
+
+    void fill_datacenter_map() {
```
Changed this for the maps to be passed to the constructor as suggested.

bitpathfinder commented on locator/tablets.cc:
```diff
@@ -700,6 +722,71 @@ class tablet_effective_replication_map : public effective_replication_map {
+            on_internal_error(tablet_logger, format("Data center: [{}] does not exist in the datacenter map.", datacenter));
+        }
+        if (dc_it->second >= dc_rf_map.size()) [[unlikely]] {
+            on_internal_error(tablet_logger, format("Data center's: [{}] index: [{}] is higher that the number of data centers in the dc_rf_map: [{}].", datacenter,
```
type's fixed

bitpathfinder commented on locator/tablets.cc:
```diff
@@ -700,6 +722,71 @@ class tablet_effective_replication_map : public effective_replication_map {
+    [[nodiscard]] size_t get_replication_factor(const dht::token id) const override {
+        auto tablet = get_tablet_map().get_tablet_id(id);
+        assert(tablet.id < _tablet_to_dc_replication_factor_map.size());
+        if (tablet.id >= _tablet_to_dc_replication_factor_map.size()) {
```
Removed asserts, leaving only checks.

bitpathfinder commented:
New version: 

- Address possible issues with added preemption point in `make_replication_map`
- Rename replication_factor_t to replication_factor_index_t 
- Change `_tablet_to_dc_replication_factor_map` from `std::vector` to `chunked_vector` to avoid large memory allocations.
- Remove unnecessary asserts in `build_dc_replication_factor_map`
- Move `build_dc_replication_factor_map` and `build_datacenter_map` out of tablets ERM, and pass these containers as constructor parameters; make `_datacenter_map` and `_tablet_to_dc_replication_factor_map`  in `tablet_effective_replication_map` immutable. 
- Fix some comments typos.
- Add `small_vector(size_t n, const T& value)` to allow construction in place and make the `small_vector` API more consistent with `std::vector`
- Change `sanity_check_read_replicas` to accept token parameter and use the new `get_replication_factor` functions in ERM.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13074/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13074/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/small_vector_test
:small_blue_diamond: topology_custom/test_change_replication_factor_1_to_0
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/162/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1618/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12719/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4691/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13074/)
#### Failed Tests (1/36333):
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13074/testReport/junit/boost.view_complex_test.test_3362_with_ttls.2.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
#### Build Details:
- Duration: 8 hr 8 min
- Builder: i-06849aeee3a7f0e22 (m5ad.8xlarge)

emdroid commented on locator/tablets.cc:
```diff
@@ -387,6 +387,60 @@ const tablet_transition_info* tablet_map::get_tablet_transition_info(tablet_id i
+    return replicas;
+}
+
+const std::optional<tablet_replica>* tablet_map::get_pending_replica(const tablet_id tablet) const {
```
do we really need to return pointer to optional in this case? as this looks a bit strange ... I know that there could be a RVO issue with the optional, but I'd prefer readability in this case - the `optional<>` already accounts for the "empty result" case and it looks like we don't need to distinguish between pointer null or optional empty (and the only place where this is used needs to do both checks then)

bitpathfinder replied:
My intention was to return the exact object that's stored in tablets, that's why I used a pointer. But agree, it might be confusing.

bitpathfinder replied:
In the new version it returns `const tablet_replica*`, to avoid unnecessary memory copying.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13098/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13098/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/small_vector_test
:small_blue_diamond: topology_custom/test_change_replication_factor_1_to_0
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/176/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12742/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4714/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1639/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13098/)
#### Build Details:
- Duration: 4 hr 13 min
- Builder: spider7.cloudius-systems.com

tgrabiec commented on locator/tablets.cc:
```diff
@@ -670,9 +670,23 @@ future<bool> check_tablet_replica_shards(const tablet_metadata& tm, host_id this
+
+public:
+    using replication_factor_index_t = uint16_t;
+    using replication_factor_list = utils::small_vector<replication_factor_index_t, 8>;
```
The element of replication_factor_list is replication factor, so the type shouldn't be `replication_factor_index_t` (dc index), it's not an index into the list, it's replication factor value.

tgrabiec commented on locator/tablets.cc:
```diff
@@ -717,15 +731,53 @@ class tablet_effective_replication_map : public effective_replication_map {
-                                     size_t replication_factor)
+                                     size_t replication_factor,
+                                     absl::flat_hash_map<sstring, replication_factor_index_t>& datacenter_map,
+                                     utils::chunked_vector<replication_factor_list>& rf_map)
```
It may be surprising that things passed by ref are moved-from. Better to capture by && and require explicit std::move() in the caller.

tgrabiec commented on locator/tablets.cc:
```diff
@@ -670,9 +670,23 @@ future<bool> check_tablet_replica_shards(const tablet_metadata& tm, host_id this
     mutable const tablet_map* _tmap = nullptr;
+    // The `_datacenter_map` stores the association between each data center
+    // and its index in the `replication_factor_list` in `_tablet_to_dc_replication_factor_map`.
+    const absl::flat_hash_map<sstring, replication_factor_index_t> _datacenter_map;
```
I think naming it `dc_index_type` would be clearer. The datacenter_map does not depend on the replication factor map, it's the other way around. It could be used for other things than replication factor. The value represents a dc, so it's more appropriate to name it dc_index or dc_id. The size of the type depends on how many DCs there can be.

tgrabiec commented on locator/tablets.cc:
```diff
@@ -717,15 +731,53 @@ class tablet_effective_replication_map : public effective_replication_map {
+
+    [[nodiscard]] size_t get_replication_factor(const dht::token id) const override {
+        auto tablet = get_tablet_map().get_tablet_id(id);
+        assert(tablet.id < _tablet_to_dc_replication_factor_map.size());
```
Should be dropped.

tgrabiec commented on locator/tablets.cc:
```diff
@@ -878,10 +930,65 @@ std::unordered_set<sstring> tablet_aware_replication_strategy::recognized_tablet
+
+// Create a simple map [data center -> index]. This index is used as position index in the small vector
+// in the tablet_to_dc_replication_factor_map.
+[[nodiscard]] absl::flat_hash_map<sstring, tablet_effective_replication_map::replication_factor_index_t> build_datacenter_map(const auto& data_centers_list) {
```
Comment needs updating, there's no tablet_to_dc_replication_factor_map. Maybe mention `replication_factor_list`.

tgrabiec commented on locator/tablets.cc:
```diff
@@ -878,10 +930,65 @@ std::unordered_set<sstring> tablet_aware_replication_strategy::recognized_tablet
+// Populate the `tablet_to_dc_replication_factor_map` with replication factors for each data center.
+// For each tablet index in the tablet map, it retrieves the current replica set using `get_endpoints_for_reading`
+// and counts replicas per data center.  Each data center's replication factor (RF) is saved under
+// its index according to `_datacenter_map`.
```
Comment needs updating, it doesn't populate tablet_to_dc_replication_factor_map.

tgrabiec commented on locator/tablets.cc:
```diff
@@ -878,10 +930,65 @@ std::unordered_set<sstring> tablet_aware_replication_strategy::recognized_tablet
+            const auto it = datacenter_map.find(datacenter);
+            if (it == datacenter_map.end()) {
+                tablet_logger.debug("Could not find datacenter: {} for node: {} ", datacenter, node);
+                continue;
```
Again, it's unsafe to continue.

bitpathfinder replied:
@tgrabiec  So, the failure occurs when the effective replication map is created before the node has been initialized. In this case, I get an empty topology, so data centers cannot be found. I'm not sure how to fix this correctly. Maybe we should skip the ERM creation until after the node has been initialized since it's unusable otherwise?

tgrabiec replied:
We should probably fix the boot order to initialize topology before initializing user tables.

bitpathfinder replied:
The issue happens because the node, that's added to topology at the time of  `load_tablet_metadata` invocation has state 'none' (I guess because it has not finished initializing yet), thus topology's dc map is not updated from this node, in the following part: 
``` C++ 
// We keep location of left nodes because they may still appear in tablet replica sets
    // and algorithms expect to know which dc they belonged to. View replica pairing needs stable
    // replica indexes.
    // But we don't consider those nodes as members of the cluster so don't update dc registry.
    if (!node->left() && !node->is_none()) {        
        const auto& dc = node->dc_rack().dc;
        const auto& rack = node->dc_rack().rack;
        const auto& endpoint = node->endpoint();
        tlogger.trace("topology[{}]: add dc {}", fmt::ptr(this), dc);
        _dc_nodes[dc].emplace(node);
        _dc_rack_nodes[dc][rack].emplace(node);
        _dc_endpoints[dc].insert(endpoint);
        _dc_racks[dc][rack].insert(endpoint);
        _datacenters.insert(dc);
    }
```

tgrabiec replied:
We shouldn't have tablet replicas on nodes which didn't finish bootstrap.

tgrabiec replied:
If the node left the cluster, it won't be present in topology's _dc_nodes etc, but its dc_rack() should still be valid.

bitpathfinder replied:
> We shouldn't have tablet replicas on nodes which didn't finish bootstrap.

This is `topology_experimental_raft/test_mv_tablets_replace` test. 
As I see tablets load the replicas from the table `system.tablets` in `read_tablet_metadata`.

bitpathfinder replied:
> We shouldn't have tablet replicas on nodes which didn't finish bootstrap.

After spending some time with our initialization logic I'd say it does not look robust (and very easy to brake). Storage service contains links and references to objects that are not initialized yet. Some functions can be used at some stages of initialization, some cannot. There is no way of learning that information other than browsing the code. Particularly `topology_state_load` depends on `_group0` and `_cdc_gens`  which are initialized much later than Tablets, and in between there are other objects initialization that depend on Tablets, so we can't just move Tablets initialization to the point where topology is initialized. 
Also, the ERMs created in `init_non_system_keyspaces` just after the Tablets initialization are not consistent since topology at that time contains only one node - the local node, whereas tablets schema contains all the replicas that were persisted in `system.tablets`.  However, it seems not really used, so it's inconsistency is not big issue.

bitpathfinder commented:
New version: 

- Make absent data center when building replication factors map an error.
- Read local node state from `system.topology` during reboot to avoid issues with absent data centers.
- Fix type names in `tablet_effective_replication_map`
- Updated function comments. 
- Using r-value references to pass `datacenter_map` and `rf_map` in  `tablet_effective_replication_map` constructor.

bitpathfinder commented:
New version:

- Rebase to the last master to fix merge conflict.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13326/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13326/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/small_vector_test
:small_blue_diamond: topology_custom/test_change_replication_factor_1_to_0
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/336/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1836/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12946/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4933/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13326/)
#### Build Details:
- Duration: 4 hr 15 min
- Builder: spider3.cloudius-systems.com

avikivity commented:
>  Since a keyspace can be represented by multiple Tablets, 

What does this mean? A keyspace isn't represented by tablets.

bitpathfinder commented:
> > Since a keyspace can be represented by multiple Tablets,
> 
> What does this mean? A keyspace isn't represented by tablets.
I meant there is one-to-many relationship between a table and Tablets, thus replication factor which is set for a keyspace does not work anymore when Tablets are in migration state.

bitpathfinder commented:
Updated description: changed `keyspace` -> `table`

bitpathfinder commented:
New version: 

- Enable replica sanity test for tablets in network topology strategy.

bitpathfinder commented:
New version: 

- remove unnecessary stack trace printing in a debug message.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13339/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13339/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/small_vector_test
:small_blue_diamond: topology_custom/test_change_replication_factor_1_to_0
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/347/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12963/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4950/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1853/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13339/)
#### Build Details:
- Duration: 4 hr 18 min
- Builder: spider1.cloudius-systems.com

dani-tweig commented:
@bitpathfinder , what is the current status and what are the next steps?
As a release blocker, it should get all the attention and help it needs

bitpathfinder commented:
> @bitpathfinder , what is the current status and what are the next steps? As a release blocker, it should get all the attention and help it needs

@dani-tweig  Currently all the review issues are resolved, waiting for review from @tgrabiec

tgrabiec commented on locator/tablets.cc:
```diff
@@ -798,15 +815,53 @@ class tablet_effective_replication_map : public effective_replication_map {
+        const auto tablet = get_tablet_map().get_tablet_id(id);
+        if (tablet.id >= _tablet_to_dc_replication_factor_map.size()) [[unlikely]] {
+            on_internal_error(tablet_logger, format("Tablet id: [{}] index is out of _tablet_to_dc_replication_factor_map, size: [{}]", tablet.id,
+                                                     _tablet_to_dc_replication_factor_map.size() - 1));
```
message says "size", but we print index of last element. Can be confusing.

bitpathfinder replied:
fixed

tgrabiec commented on locator/tablets.cc:
```diff
@@ -798,15 +815,53 @@ class tablet_effective_replication_map : public effective_replication_map {
+        }
+        if (dc_it->second >= dc_rf_map.size()) [[unlikely]] {
+            on_internal_error(tablet_logger, format("Data center's: [{}] index: [{}] is higher than the number of data centers in the dc_rf_map: [{}].",
+                                                     datacenter, dc_it->second, dc_rf_map.size()));
```
Message says "index is higher than the number", which is not quite true, since we fail also when index is equal to the number of dcs. Accurate wording would be something like "index [] is out of range, max=[]"

bitpathfinder replied:
fixed

tgrabiec commented on locator/tablets.cc:
```diff
@@ -959,10 +1014,69 @@ std::unordered_set<sstring> tablet_aware_replication_strategy::recognized_tablet
+        for (const auto& node : replicas) {
+            const locator::node* node_ptr = topo.find_node(node.host);
+            if (node_ptr == nullptr) {
+                continue;
```
This solves the problem of failing during early boot when we don't have topology, but it means we will compute wrong RF. This isn't safe.

bitpathfinder replied:
Hmm, the code in `tablet_effective_replication_map::to_replica_set` filers out the nonexistent replicas from the replica set that's returned from the `get_replicas_for_reading`, so my impression was that it's a normal situation.

gleb-cloudius replied:
With my patch series that moves storage proxy to identify nodes by ids those hosts will not be filtered.

tgrabiec replied:
> Hmm, the code in `tablet_effective_replication_map::to_replica_set` filers out the nonexistent replicas from the replica set that's returned from the `get_replicas_for_reading`, so my impression was that it's a normal situation.

For `get_replicas_for_reading()` it's safe because missing replicas will just not be contacted, same as if they are down. Dropping them for RF calculation is not safe, since we may incorrectly satisfy request with fewer replicas than is required for a quorum.

gleb-cloudius replied:
> > Options I see to fix:
> > 
> > 1. Move Tablets initialization past the `group0` init
> 
> Won't work - we need them for erm building when tables are created.

May be we can move tablet creation later, when topology is already know.

> 
> > 2. Try moving topology initialization to the earlier point.
> 
> Yes, we can load group0 snapshot earlier. That seems the cleanest.

What do you mean by "load group0 snapshot". This is what group0 initialization does. So how is it different from 1?

Also we need to consider 3 cases here:
1 Normal boot - where group0 state is available early.
2. Upgrade - where it may start to be available while everything is initialized already (but this should not be a problem for tablets since there could not be tablets during upgrade) 
3. Bootstrap. As far as I see we call init_non_system_keyspaces before even joining the cluster, so it may be completely incorrect until snapshot is transferred and loading non existing snapshot earlier will not help.

tgrabiec replied:
> > > Options I see to fix:
> > > 
> > > 1. Move Tablets initialization past the `group0` init
> > 
> > 
> > Won't work - we need them for erm building when tables are created.
> 
> May be we can move tablet creation later, when topology is already know.

That would be option 3, moving user table init past group0 init, unless we create tables without tablets, but then they're useless, and potentially unsafe, so it's still the same as option 3.

> 
> > > 2. Try moving topology initialization to the earlier point.
> > 
> > 
> > Yes, we can load group0 snapshot earlier. That seems the cleanest.
> 
> What do you mean by "load group0 snapshot". This is what group0 initialization does. So how is it different from 1?

Currently group0 init is done after messaging service init. We don't want to load user tables after messaging init, that's too late, because rpcs can already try to access tables. We could init group0 in memory state from local snapshot earlier, before user tables and before networking. This is option 2. 

> 
> Also we need to consider 3 cases here: 1 Normal boot - where group0 state is available early. 2. Upgrade - where it may start to be available while everything is initialized already (but this should not be a problem for tablets since there could not be tablets during upgrade) 3. Bootstrap. As far as I see we call init_non_system_keyspaces before even joining the cluster, so it may be completely incorrect until snapshot is transferred and loading non existing snapshot earlier will not help.

Bootstrap should be safe, since schema should be consistent with local group0 snapshot. No group0 means no user tables.

gleb-cloudius replied:
> > > > Options I see to fix:
> > > > 
> > > > 1. Move Tablets initialization past the `group0` init
> > > 
> > > 
> > > Won't work - we need them for erm building when tables are created.
> > 
> > 
> > May be we can move tablet creation later, when topology is already know.
> 
> That would be option 3, moving user table init past group0 init, unless we create tables without tablets, but then they're useless, and potentially unsafe, so it's still the same as option 3.

Currently group0 init loads everything including tablets, no?

> 
> > > > 2. Try moving topology initialization to the earlier point.
> > > 
> > > 
> > > Yes, we can load group0 snapshot earlier. That seems the cleanest.
> > 
> > 
> > What do you mean by "load group0 snapshot". This is what group0 initialization does. So how is it different from 1?
> 
> Currently group0 init is done after messaging service init. We don't want to load user tables after messaging init, that's too late, because rpcs can already try to access tables. We could init group0 in memory state from local snapshot earlier, before user tables and before networking. This is option 2.

Sometimes it is impossible to create group0 before rpc: upgrade, bootstrap. And the code needs to handle all cases.

May be what we need is to move `group0_service.setup_group0_if_exist()` before `init_non_system_keyspaces`

bitpathfinder replied:
> Currently group0 init loads everything including tablets, no?
Yes. 
I'm currently testing moved group0 init before the `init_non_system_keyspaces`. I also had to move `cdc::generation_service`  since `setup_group0_if_exist` depends on it.  One test failed:  `topology_experimental_raft/test_crash_coordinator_before_streaming`

bitpathfinder replied:
Since tablets are initialized during group0 snapshot loading, probably we could remove the code below 
``` C++ 
            try {
                ss.local().load_tablet_metadata({}).get();
            } catch (...) {
                if (!cfg->maintenance_mode()) {
                    throw;
                }
                startlog.error("Failed to load tablet metadata (ignoring due to maintenance mode): {}", std::current_exception());
            }
```

bitpathfinder replied:
It looks like we need another PR, that will take care of the Tablets-Topology inconsistency in the ERM  in different cases.  For instance, when a node is bootstrapped and  a new snapshot is loaded in `group0_state_machine::transfer_snapshot` Tablets are updated earlier in the `schema_applier::merge_tables_and_views", and then in `add_column_family` a new ERM is created, that contains new Tablets map and old topology.

tgrabiec replied:
> It looks like we need another PR, that will take care of the Tablets-Topology inconsistency in the ERM in different cases. For instance, when a node is bootstrapped and a new snapshot is loaded in `group0_state_machine::transfer_snapshot` Tablets are updated earlier in the `schema_applier::merge_tables_and_views", and then in `add_column_family` a new ERM is created, that contains new Tablets map and old topology.

We should fix that first, otherwise this PR will introduce a regression where we would fail to construct ERM in this case.

bitpathfinder replied:
> We should fix that first, otherwise this PR will introduce a regression where we would fail to construct ERM in this case.


I see it as 2 different issues:

1.  Incorrect Replication Factor calculation when tablets are in the transition state, that is addressed by this PR
2.  Occasionally invalid Effective Replication Map, containing Tablets map inconsistent with Topology. This PR does not introduce this issue, it just reveals an already existing issue. Tablets metadata should be updated only together with topology, and before creating the new ERM. 

So as these are 2 different problems IMO these should be fixed separately. This PR would dependent on the other PR.  @kbr-scylla mentioned that there is a PR https://github.com/scylladb/scylladb/pull/20853, that be solving the second issue. I'm looking at it now.

bitpathfinder replied:
Created https://github.com/scylladb/scylladb/issues/21856

tgrabiec replied:
> So as these are 2 different problems IMO these should be fixed separately. This PR would dependent on the other PR. @kbr-scylla mentioned that there is a PR https://github.com/scylladb/scylladb/pull/20853

Ok.

tgrabiec commented on locator/tablets.cc:
```diff
@@ -959,10 +1014,69 @@ std::unordered_set<sstring> tablet_aware_replication_strategy::recognized_tablet
+                tablet_logger.info("topology [{}] DC map: {}", fmt::ptr(&topo),  topo.get_datacenters());
+
+                on_internal_error(tablet_logger, format("Could not find datacenter: {} for node: {} ", datacenter, node));
+                //continue;
```
Leftover comment-out.

bitpathfinder replied:
fixed

bitpathfinder commented on main.cc:
```diff
@@ -594,20 +594,47 @@ void dump_performance_profiles() {
         // still be found in the config.
         tm.get_topology().set_host_id_cfg(host_id);
-        tm.get_topology().add_or_update_endpoint(host_id, endpoint);
+        tm.get_topology().add_or_update_endpoint(host_id, endpoint, std::nullopt, node_state);
```
This patch does not insure that.  The ERM created at that point contains only the local node which is apparently required for further initialization. Then later when raft server starts it loads the topology from the system tables and a new ERM is created.

tgrabiec replied:
Then I'm not sure what you mean in the commit log which says "This ensures correct initialization of ... data center list in topology". It probably needs to be revised.

bitpathfinder replied:
This change was removed since topology initialization during reboot is now done earlier and takes care of initializing the local node state in case the node was bootstrapped.

bitpathfinder commented on service/raft/raft_group0_client.cc:
```diff
@@ -33,6 +33,30 @@ namespace service {
+    if (!s || *s == "use_pre_raft_procedures") {
+        return service::group0_upgrade_state::use_pre_raft_procedures;
+    }
+    else if (*s == "synchronize") {
```
fixed

bitpathfinder committed:
Fix logging: node ptr `nptr` is printed after move in
`topology::add_node`

In add_node the trace log was incorrectly printing the added node, since
`nptr` was no more valid after move at line 185.

bhalevy commented:
@bitpathfinder:
1. the series needs rebasing
2. Can you please added "Fixes #20282" to the cover letter, since it's blocked on this PR too.

bitpathfinder commented:
> @bitpathfinder:
> 
>     1. the series needs rebasing
> 
>     2. Can you please add "Fixes [(tablets only) All DC nodes got Coredump and error of: "failed to log message: fmt='Requested location for node {} not in topology" on altering keyspace replication-factor of one data-center (out of 2)  #20282](https://github.com/scylladb/scylladb/issues/20282)" to the cover letter, since it's blocked on this PR too.

@bhalevy Sure, I will rebase, but there are still changes pending to fix one review issue, so my intention was to preform the rebasing after the fix, so it will be easier to track the changes between versions.

bitpathfinder commented:
Description updated: added `fixes scylladb/scylladb#20282`

bitpathfinder committed:
Init: load topology before non system keyspace init.

To ensure ERM consistency, the topology initialization has been
moved to occur before the creation of tablet-based ERMs during
the initialization of non-system keyspaces. Since topology is
initialized during the Group 0 snapshot loading process, the
Group0 initialization has been shifted to take place before
the `init_non_system_keyspaces`.

bitpathfinder commented:
New version: 

- Init: moved  group0 snapshot loading before the `init_non_system_keyspaces` to insure proper topology initialization.
- Removed commits that became unnecessary.

bitpathfinder commented:
New version: 

- Remove unnecessary log. 
- Make absent node an error.

bitpathfinder commented:
As discussed with @kbr-scylla changing to P2, since the node crash is fixed in https://github.com/scylladb/scylladb/pull/20851.

mykaul commented:
Wow, loads of conflicts :-/  - please fix.

bitpathfinder commented:
> Wow, loads of conflicts :-/ - please fix.

Already done locally.  Will publish in another PR soon.

bitpathfinder commented:
New version: 

- Rebase to the latest master.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13717/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13717/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/small_vector_test
:small_blue_diamond: topology_custom/test_change_replication_factor_1_to_0
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/593/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5263/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13276/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2163/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13717/)
#### Failed Tests (935/36791):
- [test_rolling_upgrade](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13717/testReport/junit/rolling_upgrade_test/TestRollingUpgrade/Tests___dtest_with_topology_changes___test_rolling_upgrade) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_rolling_upgrade)
- [test_simple_removenode_4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13717/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_removenode_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_removenode_4)
- [test_simple_removenode_4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13717/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_removenode_4_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_removenode_4)
- [test_rolling_upgrade](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13717/testReport/junit/rolling_upgrade_test/TestRollingUpgrade/Tests___dtest___test_rolling_upgrade) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_rolling_upgrade)
- [test_consistent_reads_after_bootstrap](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13717/testReport/junit/consistent_bootstrap_test/TestBootstrapConsistency/Tests___dtest_with_tablets___test_consistent_reads_after_bootstrap) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_consistent_reads_after_bootstrap)
- [test_consistent_reads_after_bootstrap](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13717/testReport/junit/consistent_bootstrap_test/TestBootstrapConsistency/Tests___dtest_with_tablets___test_consistent_reads_after_bootstrap_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_consistent_reads_after_bootstrap)
- [test_cluster_cleanup_no_resurrection](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13717/testReport/junit/cleanup_test/TestCleanup/Tests___dtest_with_tablets___test_cluster_cleanup_no_resurrection) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cluster_cleanup_no_resurrection)
- [test_restore_snapshot_using_old_token_ownership](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13717/testReport/junit/backup_restore_tests/TestBackupRestore/Tests___dtest_with_tablets___test_restore_snapshot_using_old_token_ownership) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_snapshot_using_old_token_ownership)
- [test_cluster_cleanup_no_resurrection](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13717/testReport/junit/cleanup_test/TestCleanup/Tests___dtest_with_tablets___test_cluster_cleanup_no_resurrection_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cluster_cleanup_no_resurrection)
- [test_restore_snapshot_using_old_token_ownership](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13717/testReport/junit/backup_restore_tests/TestBackupRestore/Tests___dtest_with_tablets___test_restore_snapshot_using_old_token_ownership_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_snapshot_using_old_token_ownership)
#### Build Details:
- Duration: 8 hr 12 min
- Builder: i-00041ad359a60777e (m5ad.8xlarge)

bitpathfinder commented:
Blocked on https://github.com/scylladb/scylladb/issues/21856

bhalevy commented:
Based on https://github.com/scylladb/scylladb/pull/21289#issuecomment-2518015641 removing the `release blocker` label as well.

tgrabiec commented on main.cc:
```diff
@@ -1718,14 +1718,34 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
+             * so CDC generation service takes sharded<db::sys_dist_ks> and must check local_is_initialized()
+             * every time it accesses it (because it may have been stopped already), then take local_shared()
+             * which will prevent sys_dist_ks from being destroyed while the service operates on it.
+             */
```
Interestingly, this comment was wrong before this change, sys_disk_ks was stopped after CDC generation service.

Btw, how is setup_group0_if_exists() depending on cdc_generation_service? It's not used explicitly.

bitpathfinder replied:
This code in `storage_service::topology_state_load(state_change_hint hint)` requires it. 
``` C++
for (const auto& gen_id : _topology_state_machine._topology.committed_cdc_generations) {
        rtlogger.trace("topology_state_load: process committed cdc generation {}", gen_id);
>>        co_await _cdc_gens.local().handle_cdc_generation(gen_id);
        if (gen_id == _topology_state_machine._topology.committed_cdc_generations.back()) {
            co_await _sys_ks.local().update_cdc_generation_id(gen_id);
            rtlogger.debug("topology_state_load: the last committed CDC generation ID: {}", gen_id);
        }
    }
```

bitpathfinder commented:
New version: rebase and fix merge conflicts.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14040/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14040/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/small_vector_test
:small_blue_diamond: topology_custom/test_change_replication_factor_1_to_0
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/858/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5601/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13614/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/262/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2484/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14040/)
#### Failed Tests (1054/37085):
- [test_rolling_upgrade](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14040/testReport/junit/rolling_upgrade_test/TestRollingUpgrade/Tests___dtest___test_rolling_upgrade) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_rolling_upgrade)
- [test_consistent_reads_after_bootstrap](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14040/testReport/junit/consistent_bootstrap_test/TestBootstrapConsistency/Tests___dtest_with_tablets___test_consistent_reads_after_bootstrap) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_consistent_reads_after_bootstrap)
- [test_consistent_reads_after_bootstrap](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14040/testReport/junit/consistent_bootstrap_test/TestBootstrapConsistency/Tests___dtest_with_tablets___test_consistent_reads_after_bootstrap_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_consistent_reads_after_bootstrap)
- [test_cluster_cleanup_no_resurrection](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14040/testReport/junit/cleanup_test/TestCleanup/Tests___dtest_with_tablets___test_cluster_cleanup_no_resurrection) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cluster_cleanup_no_resurrection)
- [test_restore_snapshot_using_old_token_ownership](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14040/testReport/junit/backup_restore_tests/TestBackupRestore/Tests___dtest_with_tablets___test_restore_snapshot_using_old_token_ownership) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_snapshot_using_old_token_ownership)
- [test_cluster_cleanup_no_resurrection](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14040/testReport/junit/cleanup_test/TestCleanup/Tests___dtest_with_tablets___test_cluster_cleanup_no_resurrection_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cluster_cleanup_no_resurrection)
- [test_restore_snapshot_using_old_token_ownership](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14040/testReport/junit/backup_restore_tests/TestBackupRestore/Tests___dtest_with_tablets___test_restore_snapshot_using_old_token_ownership_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_snapshot_using_old_token_ownership)
- [test_simple_bootstrap](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14040/testReport/junit/bootstrap_test/TestBootstrap/Tests___dtest_with_tablets___test_simple_bootstrap) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_bootstrap)
- [test_tablets_data_is_segregated_after_bootstrap](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14040/testReport/junit/compaction_additional_test/TestTimeWindowDataSegregation/Tests___dtest_with_tablets___test_tablets_data_is_segregated_after_bootstrap) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablets_data_is_segregated_after_bootstrap)
- [test_simple_bootstrap](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14040/testReport/junit/bootstrap_test/TestBootstrap/Tests___dtest_with_tablets___test_simple_bootstrap_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_bootstrap)
#### Build Details:
- Duration: 4 hr 3 min
- Builder: spider7.cloudius-systems.com

bitpathfinder commented:
New version: 

* Rebase and fix conflicts.

scylladb-promoter commented:
#### :red_circle: CI State: ABORTED

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14212/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14212/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/small_vector_test
:small_blue_diamond: topology_custom/test_change_replication_factor_1_to_0
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5717/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13730/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2604/)
:x: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/976/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/380/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14212/)
#### Failed Tests (1535/36346):
- [test_consistent_reads_after_bootstrap](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14212/testReport/junit/consistent_bootstrap_test/TestBootstrapConsistency/Tests___dtest_with_tablets___test_consistent_reads_after_bootstrap) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_consistent_reads_after_bootstrap)
- [test_consistent_reads_after_bootstrap](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14212/testReport/junit/consistent_bootstrap_test/TestBootstrapConsistency/Tests___dtest_with_tablets___test_consistent_reads_after_bootstrap_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_consistent_reads_after_bootstrap)
- [test_cluster_cleanup_no_resurrection](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14212/testReport/junit/cleanup_test/TestCleanup/Tests___dtest_with_tablets___test_cluster_cleanup_no_resurrection) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cluster_cleanup_no_resurrection)
- [test_cluster_cleanup_no_resurrection](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14212/testReport/junit/cleanup_test/TestCleanup/Tests___dtest_with_tablets___test_cluster_cleanup_no_resurrection_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cluster_cleanup_no_resurrection)
- [test_restore_snapshot_using_old_token_ownership](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14212/testReport/junit/backup_restore_tests/TestBackupRestore/Tests___dtest_with_tablets___test_restore_snapshot_using_old_token_ownership) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_snapshot_using_old_token_ownership)
- [test_restore_snapshot_using_old_token_ownership](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14212/testReport/junit/backup_restore_tests/TestBackupRestore/Tests___dtest_with_tablets___test_restore_snapshot_using_old_token_ownership_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_snapshot_using_old_token_ownership)
- [test_simple_bootstrap](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14212/testReport/junit/bootstrap_test/TestBootstrap/Tests___dtest_with_tablets___test_simple_bootstrap) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_bootstrap)
- [test_tablets_data_is_segregated_after_bootstrap](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14212/testReport/junit/compaction_additional_test/TestTimeWindowDataSegregation/Tests___dtest_with_tablets___test_tablets_data_is_segregated_after_bootstrap) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablets_data_is_segregated_after_bootstrap)
- [test_simple_bootstrap](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14212/testReport/junit/bootstrap_test/TestBootstrap/Tests___dtest_with_tablets___test_simple_bootstrap_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_bootstrap)
- [test_manual_bootstrap](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14212/testReport/junit/bootstrap_test/TestBootstrap/Tests___dtest_with_tablets___test_manual_bootstrap) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_manual_bootstrap)
#### Build Details:
- Duration: 12 hr
- Builder: i-0dab6233f96d3b430 (m5ad.12xlarge)

mykaul commented:
@bitpathfinder - what's the status of this PR?

bitpathfinder commented:
> @bitpathfinder - what's the status of this PR?

This PR waits for scylladb/scylladb#20853.

mykaul commented:
Lots of conflicts...

mykaul commented:
Goziilion of conflicts. Is this PR being worked on?

bitpathfinder commented:
> > > Goziilion of conflicts. Is this PR being worked on?
> > 
> > 
> > No, it's currently on hold, waiting for #25302 (which completes work started in #20853)
> 
> @bitpathfinder the above was merged.

I know, but all the focus now is on the strong consistency. 
To complete this PR first it will be needed to fix possible topology/tablets inconsistency https://github.com/scylladb/scylladb/issues/21856 (which I was hoping would be fixed by #25302)

bhalevy commented:
Based on @tgrabiec's [comment](https://github.com/scylladb/scylladb/pull/24421#discussion_r2985024899) we need this fixed.
@bitpathfinder @calebxyz should we reassign to someone else to pick up this PR and complete it?
